### DISCUSSION TITLE
GA4 support With Coreshop 2.2

### DISCUSSION THREAD

**Discussion by rushitkhunt1996** [coreshop/CoreShop#2324] [Q&A]
Hi @dpfaffenbauer and @All,

We used coreshop 2.2 version with our old eCommerce website at that time we used UA(Universal Analytics) for tracking, now we want to upgrade it with GA4, so we are not able to find any configuration for GA4.
Also, we found one issue raised with GA4 but it's for Coreshop 3.1. [click here](https://github.com/coreshop/CoreShop/pull/2303)

Anyone can guide what we need to do for GA4 support on coreshop 2.2.

**Comment by dpfaffenbauer** [comment]
sort of like this: https://github.com/coreshop/CoreShop/pull/2303

**Reply by divyashah2018** [reply]
1. features/tracking/analytics_enhanced.feature -  ga('require', 'ec'); ga('set', 'currencyCode', 'EUR'); ga('ec:addImpression', {"id":##id##,"name":"T-Shirt","category":"","price":24,"quantity":1,"currency":"EUR"});
2. features/tracking/ga4.feature- new file added
3. features/tracking/ga4_tag_manager.feature- new file added
4. features/domain/tracking/matomo.feature- removed
5. features/tracking/tag_manager_enhanced.feature- multiple changes according to the link-https://github.com/coreshop/CoreShop/pull/2303/files#diff-bc27cd4e4dc3d7131393a86fb8211617e534c36498ee40c1b459d0e51f9b25fd
6. src/CoreShop/Behat/Context/Domain/TrackingContext.php- multiple changes according to the link-https://github.com/coreshop/CoreShop/pull/2303/files#diff-bc27cd4e4dc3d7131393a86fb8211617e534c36498ee40c1b459d0e51f9b25fd
7. src/CoreShop/Behat/Resources/config/suites/domain/menu.yml-  domain_tracking: to domain_menu:
8. src/CoreShop/Behat/Service/Tracking/ConfigResolver.php- added  
                                                        class ConfigResolver implements ConfigResolverInterface
                                                        {

                                                            public function __construct(private bool $addGtagCode = false)
                                                            {
                                                            }

                                                            public function getGoogleConfig(): ConfigObject
                                                            {
                                                                $params = [];

                                                                if ($this->addGtagCode) {
                                                                    $params['gtagcode'] = 'coreshop';
                                                                }

                                                                /**
                                                                * @psalm-suppress DeprecatedClass
                                                                */
                                                                return new ConfigObject([]);
                                                                return new ConfigObject($params);

9. src/CoreShop/Bundle/TrackingBundle/Resources/config/services.yml - Added
                                                        CoreShop\Bundle\TrackingBundle\Tracker\Google\GA4Ecommerce:
                                                                parent: CoreShop\Bundle\TrackingBundle\Tracker\AbstractEcommerceTracker
                                                                public: true
                                                                calls:
                                                                    - [setTracker, ['@Pimcore\Analytics\Google\Tracker']]
                                                                    - [setConfigResolver, ['@CoreShop\Bundle\TrackingBundle\Resolver\ConfigResolver']]
                                                                tags:
                                                                    - { name: coreshop.tracking.tracker, type: google-analytics-4 }

                                                            CoreShop\Bundle\TrackingBundle\Tracker\Google\GA4TagManagerEcommerce:
                                                                parent: CoreShop\Bundle\TrackingBundle\Tracker\AbstractEcommerceTracker
                                                                public: true
                                                                calls:
                                                                    - [setTracker, ['@Pimcore\Analytics\Google\Tracker']]
                                                                    - [setConfigResolver, ['@CoreShop\Bundle\TrackingBundle\Resolver\ConfigResolver']]
                                                                tags:
                                                                    - { name: coreshop.tracking.tracker, type: google-analytics-4-tag-manager }
                                                            
10. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4- ga4 folder added
11. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4/checkout.js.twig- added gtag('event', '{{ event }}', {{ actionData|json_encode()|raw }});
12. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4/checkout_complete.js.twig- added gtag('event', 'purchase', {{ actionData|json_encode()|raw }});
13. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4/product_action.js.twig- added gtag('event', '{{ event }}', {{ actionData|json_encode()|raw }});
14. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4/product_impression.js.twig-added gtag('event', 'view_item_list', {{ actionData|json_encode()|raw }});
15. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4/product_view.js.twig- added gtag('event', 'view_item', {{ actionData|json_encode()|raw }});
16. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm- ga4tm folder added
17. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm/checkout.js.twig- added dataLayer.push({ ecommerce: null });
                                                dataLayer.push({'event': '{{ event }}', ecommerce: {{ actionData|json_encode()|raw }}});
18. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm/checkout_complete.js.twig - added dataLayer.push({ ecommerce: null });
dataLayer.push({'event': 'purchase', ecommerce: {{ actionData|json_encode()|raw }}});
19. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm/product_action.js.twig - added dataLayer.push({ ecommerce: null });
dataLayer.push({'event': '{{ event }}', ecommerce: {{ actionData|json_encode()|raw }}});
20. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm/product_impression.js.twig - added dataLayer.push({ ecommerce: null });
dataLayer.push({'event': 'view_item_list', ecommerce: {{ actionData|json_encode()|raw }}});
21. src/CoreShop/Bundle/TrackingBundle/Resources/views/Tracking/ga4tm/product_view.js.twig - added dataLayer.push({ ecommerce: null });
dataLayer.push({'event': 'view_item', ecommerce: {{ actionData|json_encode()|raw }}});
22. src/CoreShop/Bundle/TrackingBundle/Tracker/Google/AnalyticsEnhancedEcommerce.php - edited in transformProductAction function
             'brand' => $item['brand'] ?? null,
            'variant' => $item['variant'] ?? null,
            'quantity' => $item['quantity'] ?? 1,
            'position' => $item['position'] ?? null,
23. src/CoreShop/Bundle/TrackingBundle/Tracker/Google/GA4Ecommerce.php - new file added https://github.com/coreshop/CoreShop/pull/2303/files#diff-bc27cd4e4dc3d7131393a86fb8211617e534c36498ee40c1b459d0e51f9b25fd
24. src/CoreShop/Bundle/TrackingBundle/Tracker/Google/GA4TagManagerEcommerce.php - new file added https://github.com/coreshop/CoreShop/pull/2303/files#diff-bc27cd4e4dc3d7131393a86fb8211617e534c36498ee40c1b459d0e51f9b25fd
25. src/CoreShop/Bundle/TrackingBundle/Tracker/Google/UniversalEcommerce.php - edited in transformProductAction function
             'brand' => $item['brand'] ?? null,
            'variant' => $item['variant'] ?? null,
            'quantity' => $item['quantity'] ?? 1,
            'position' => $item['position'] ?? null,
            
     Tried this changes on the Coreshop 2.2 but not working

**Reply by dpfaffenbauer** [reply]
you have to implement it as an extension sort of CoreShop implemented it in 3.1

**Reply by rushitkhunt1996** [reply]
Hi @dpfaffenbauer,

Is this working with Coreshop 3.0?, because we tried to add the script on Coreshop 3.0 but they're also not working.
Also, we check code of https://demo3.coreshop.org/ on the live demo, there also GA4 is not implemented.

Can you please guide us in detail so we can take some action on it, it's a massive blocker for us.

**Reply by dpfaffenbauer** [reply]
It has been released with CoreShop 3.1. It is not activated on the demo. Here you can find the documentation for trackers and how to add a custom one: https://docs.coreshop.org/3.1.0/Development/Ecommerce_Tracking/index.html
